src/Razor/src/Microsoft.AspNetCore.Razor.ProjectEngineHost/Utilities/AsyncBatchingWorkQueue`2.cs

```diff
@@ -31,6 +31,15 @@ namespace Microsoft.AspNetCore.Razor.Utilities;
 /// </summary>
 internal class AsyncBatchingWorkQueue<TItem, TResult>
 {
+    /// <summary>
+    /// Fired when all batches have finished being processed, and the queue is waiting for an AddWork call.
+    /// </summary>
+    /// <remarks>
+    /// This is a best-effort signal with no guarantee that more work won't be queued, and hence the queue
+    /// going non-idle, immediately after (or during!) the event firing.
+    /// </remarks>
+    public event EventHandler? Idle;
```

Comment:
Why is this an event? That seems very different from the rest of the `AsyncBatchingWorkQueue` API which takes all relevant values in the constructor rather than through settable properties. Also, why was `EventHandler` chosen? Are the sender and `EventArgs` useful?

Consider just adding a constructor overload that takes an `Action` parameter if there aren't answers to those questions. I'd prefer to see a readonly `Action` field rather than a new mutable event field added to a type that is already so careful about locking semantics and mutable fields.

Reply:
I did start down the path of doing this as a constructor parameter, but it felt like a little bit of overkill for something that I didn't expect many consumers to provide. An event seemed better for representing something that is optional, rarely used, and doesn't affect the core logic and purpose of the queue.

Having said that, I don't feel very strongly about it, so I'm happy to change this to an `Action` per your suggestion

Reply:
I'd just add a constructor overload so that most consumers can ignore it.